tests/test_bootstrap.py

```diff
@@ -0,0 +1,551 @@
+    between all bootstraps which inherits from class
+    :class:`~pythonforandroid.bootstrap.Bootstrap`.
+    """
+
+    def test_properties(self):
+        """A test which will initialize a bootstrap and will check if the
+        values are the expected.
+        """
+        bs = Bootstrap().get_bootstrap("sdl2", self.ctx)
+        self.assertEqual(bs.name, "sdl2")
+        self.assertEqual(bs.jni_dir, "sdl2/jni")
+        self.assertEqual(bs.get_build_dir_name(), "sdl2-python3")
+
+        # test dist_dir error
+        with self.assertRaises(SystemExit) as e:
+            getattr(bs, "dist_dir")
+        self.assertEqual(e.exception.args[0], 1)
+
+        # test dist_dir success
+        self.setUp_distribution_with_bootstrap(bs)
+        self.assertTrue(bs.dist_dir.endswith("dists/test_prj"))
+
+    def test_build_dist_dirs(self):
+        """A test which will initialize a bootstrap and will check if the
+        directories we set has the values that we expect. Her we test methods:
+
+            - :meth:`~pythonforandroid.bootstrap.Bootstrap.get_build_dir`
+            - :meth:`~pythonforandroid.bootstrap.Bootstrap.get_dist_dir`
+            - :meth:`~pythonforandroid.bootstrap.Bootstrap.get_common_dir`
+        """
+        bs = Bootstrap().get_bootstrap("sdl2", self.ctx)
+        self.setUp_distribution_with_bootstrap(bs)
+
+        self.assertTrue(
+            bs.get_build_dir().endswith("build/bootstrap_builds/sdl2-python3")
+        )
+        self.assertTrue(bs.get_dist_dir("test_prj").endswith("dists/test_prj"))
+        self.assertTrue(
+            bs.get_common_dir().endswith("pythonforandroid/bootstraps/common")
+        )
+
+    def test_list_bootstraps(self):
+        """A test which will initialize a bootstrap and will check if the
+        method :meth:`~pythonforandroid.bootstrap.Bootstrap.list_bootstraps`
+        returns the expected values, which should be: `empty", `service_only`,
+        `webview` and `sdl2`
+        """
+        # prepare bootstrap and distribution
+        bs = Bootstrap().get_bootstrap("sdl2", self.ctx)
+        self.setUp_distribution_with_bootstrap(bs)
+
+        expected_bootstraps = {"empty", "service_only", "webview", "sdl2"}
+        set_of_bootstraps = set(Bootstrap().list_bootstraps())
+        self.assertEqual(
+            expected_bootstraps, expected_bootstraps & set_of_bootstraps
+        )
+        self.assertEqual(len(expected_bootstraps), len(set_of_bootstraps))
+
+    def test_get_bootstraps_from_recipes(self):
+        """A test which will initialize a bootstrap and will check if the
+        method :meth:`~pythonforandroid.bootstrap.Bootstrap.
+        get_bootstraps_from_recipes` returns the expected values
+        """
+        recipes_service = {"pyjnius", "python3"}
+        bs = Bootstrap().get_bootstrap_from_recipes(recipes_service, self.ctx)
+        self.setUp_distribution_with_bootstrap(bs)
+
+        self.assertEqual(bs.name, "service_only")
```

Comment:
This guy seems to be returning `webview` on my setup for some reason :thinking: 
```
[andre:~/workspace/python-for-android] develop+* 1 ± venv/bin/pytest tests/test_bootstrap.py 
================================================================================================================================== test session starts ==================================================================================================================================
platform linux -- Python 3.6.7, pytest-4.2.0, py-1.7.0, pluggy-0.8.1
rootdir: /home/andre/workspace/python-for-android, inifile:
collected 23 items                                                                                                                                                                                                                                                                      

tests/test_bootstrap.py ...F...................                                                                                                                                                                                                                                   [100%]

======================================================================================================================================= FAILURES ========================================================================================================================================
__________________________________________________________________________________________________________________ TestBootstrapBasic.test_get_bootstraps_from_recipes __________________________________________________________________________________________________________________

self = <test_bootstrap.TestBootstrapBasic testMethod=test_get_bootstraps_from_recipes>

    def test_get_bootstraps_from_recipes(self):
        """A test which will initialize a bootstrap and will check if the
        method :meth:`~pythonforandroid.bootstrap.Bootstrap.
        get_bootstraps_from_recipes` returns the expected values
        """
        recipes_service = {"pyjnius", "python3"}
        bs = Bootstrap().get_bootstrap_from_recipes(recipes_service, self.ctx)
        self.setUp_distribution_with_bootstrap(bs)
    
>       self.assertEqual(bs.name, "service_only")
E       AssertionError: 'webview' != 'service_only'
E       - webview
E       + service_only

tests/test_bootstrap.py:134: AssertionError
--------------------------------------------------------------------------------------------------------------------------------- Captured stderr call ----------------------------------------------------------------------------------------------------------------------------------
[INFO]:    Trying to find a bootstrap that matches the given recipes.
[INFO]:    Found 3 acceptable bootstraps: ['webview', 'sdl2', 'service_only']
[INFO]:    Using the first of these: webview
[INFO]:    No existing dists meet the given requirements!
----------------------------------------------------------------------------------------------------------------------------------- Captured log call -----------------------------------------------------------------------------------------------------------------------------------
bootstrap.py               156 INFO     [INFO]:    Trying to find a bootstrap that matches the given recipes.
bootstrap.py               186 INFO     [INFO]:    Found 3 acceptable bootstraps: ['webview', 'sdl2', 'service_only']
bootstrap.py               189 INFO     [INFO]:    Using the first of these: webview
distribution.py            111 INFO     [INFO]:    No existing dists meet the given requirements!
=================================================================================================================================== warnings summary ====================================================================================================================================
pythonforandroid/util.py:23
  /home/andre/workspace/python-for-android/pythonforandroid/util.py:23: DeprecationWarning: WgetDownloader style of invoking requests is deprecated. Use newer urlopen functions/methods
    urlretrieve = WgetDownloader().retrieve

-- Docs: https://docs.pytest.org/en/latest/warnings.html
==================================================================================================================== 1 failed, 22 passed, 1 warnings in 0.31 seconds ====================================================================================================================
[andre:~/workspace/python-for-android] develop+* 1 ±
```

Reply:
I changed the recipes of this tests in such a way that we could only get only one result so we can avoid this issue (I suspect that we got different results because we use `sets` in some calculations and probably we convert it to a list at some point...anyway...the result that you got I consider it valid...so I think it's better to avoid these ambiguities).
**Thanks!!**

Reply:
Thanks for the update, then it means you probably discovered a bug in this component :clap: 
Basically https://github.com/kivy/python-for-android/blob/v2019.06.06/pythonforandroid/bootstrap.py#L190
So at some point we assume something is ordered but it's not. I think the problem may actually come from us assuming the [listdir()](https://github.com/kivy/python-for-android/blob/v2019.06.06/pythonforandroid/bootstrap.py#L145) of [list_bootstraps()](https://github.com/kivy/python-for-android/blob/v2019.06.06/pythonforandroid/bootstrap.py#L141) is in a known/consistent order while it's not.
I've created https://github.com/kivy/python-for-android/issues/1875 to investigate, let me know what you think

Reply:
You are right, We could have a problem with this because we don't have any control of what bootstrap will be returned in case that there are multiple possibilities...also I think that this bootstraps list should not been sorted alphabetically instead should be done based on requested recipes...but better to discuss it in the issue you created

**Note:** I cannot take the credits  of discover this bug, actually we both found it...you pointed exactly what is the problem and your local testing triggered the the alarm  :clap: